### PR TITLE
Fixed the assertEquals for two dimensional arrays

### PR DESCRIPTION
**Pull request by atulagrawal**
Fixes # .https://github.com/cbeust/testng/issues/1661

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by atulagrawal** [comment]
@krmahadevan @juherr please review and let me know your feedback.

**Comment by juherr** [comment]
@atulagrawal the change breaks the build. Check if you don't have a regression with some other tests.

**Comment by atulagrawal** [comment]
@juherr - There are following two types of comparison for arrays as per thread in #790 and #870 :
1. compare by value - is done for an standalone array (one dimensional)
2. compare by reference - is done when an arrays that is contained by collection, set, map etc and for multi-dimensional array.
The reason for #1661 is because multiple dimensional arrays are converted to collection is are compared via reference.
In order to fix this issue, either I will have to add custom logic in assert method for handling multi-dimensional arrays or will have to change the behavior of collections comparison.

**Comment by atulagrawal** [comment]
IMHO if someone asserts a collection or any nested data structure, they will look for value comarision. In most cases, developers make expected structure while the method under test returns the actual structure. On top of this, the value comparison covers reference comparision but reverse is not true.

**Comment by juherr** [comment]
For value comparison in collections, you have `assertEqualsDeep`.

**Comment by atulagrawal** [comment]
@juherr I have updated the logic of public static void assertEquals(Object[] actual, Object[] expected, String message)  method along with unit tests cases.

**Comment by krmahadevan** [comment]
@atulagrawal 

> This resulted in compare-by-reference behavior.

Are you sure about this ?

Here's the call chain I am seeing:

1. `org.testng.Assert#assertEquals(java.lang.Object[], java.lang.Object[])` invokes 
2. `org.testng.Assert#assertEquals(java.lang.Object[], java.lang.Object[], java.lang.String)` invokes 
3. `org.testng.Assert#assertEquals(java.util.Collection<?>, java.util.Collection<?>, java.lang.String)`  which iterates through each of the elements and invokes 
4. `org.testng.Assert#assertEqualsImpl` (where `equals()` comparison is being done.

Where is the reference comparison being done here? Can you please elaborate?

File: src/test/java/test/asserttests/ArrayEqualityAssertTest.java [hunk 1]
```diff
@@ -281,15 +281,15 @@ public void arrayDeepInIterablesAssertEquals() {
         "arrays inside Iterables which are inside Iterables themselves are compared by reference in assertEquals");
   }
 
-  @Test(expectedExceptions = AssertionError.class)
+  @Test
   public void arrayInsideArrayAssertEquals() {
     int[][] array = new int[][] {new int[] {42}};
     int[][] arrayCopy = new int[][] {new int[] {42}};
     assertEquals(
         array, arrayCopy, "arrays inside arrays are compared by reference in assertEquals");
```
Comment (juherr) [inline src/test/java/test/asserttests/ArrayEqualityAssertTest.java]: `by value`

File: src/main/java/org/testng/Assert.java [hunk 2]
```diff
@@ -1077,7 +1076,24 @@ public static void assertEquals(Object[] actual, Object[] expected, String messa
         fail("Arrays not equal: " + Arrays.toString(expected) + " and " + Arrays.toString(actual));
       }
     }
-    assertEquals(Arrays.asList(actual), Arrays.asList(expected), message);
+    if (actual.length != expected.length) {
```
Comment (juherr) [inline src/main/java/org/testng/Assert.java]: What if we add a `public static void assertEquals(Object[][] actual, Object[][] expected, String message)`?
Does it work?

Comment (atulagrawal) [reply src/main/java/org/testng/Assert.java]: No. Its not working this way.

File: src/test/java/test/asserttests/ArrayEqualityAssertTest.java (hunk 1 above)
Comment (atulagrawal) [reply src/test/java/test/asserttests/ArrayEqualityAssertTest.java]: Changed to value.

**Review by juherr — APPROVED** [review]
LGTM 👍 

Ping @cbeust if you agree with the behavior change

**Pull request merged**